File: aws-ec2/operation/create_application_load_balancer.rb
require 'json'
@log.trace("Started executing 'fb-cloud:aws-ec2:operation:create_application_load_balancer.rb' flintbit...")
# Flintbit Input Parameters
# Mandatory
connector_name = 'amazon-ec2'	      # Name of the Amazon EC2 Connector
action = "create-application-load-balancer"  # Specifies the name of the operation: create-security-group
load_balancer_name = @input.get('name')
listener_array = @input.get('listeners')       # JSONArray of listners
subnet_array = @input.get('subnets')           # Array of subnets on which we want to connect load balancer
region = @input.get('region')
scheme = @input.get('scheme')
ip_address_type = @input.get('ip-address-type')
# Optional
@access_key = @input.get('access-key')
@secret_key = @input.get('security-key')
request_timeout = @input.get('timeout')# Execution time of the Flintbit in milliseconds (default timeout is 60000 milloseconds)
security_groups = @input.get('security-groups')
tags = @input.get('tags')


@log.info("Flintbit input parameters are, action : #{action} |  Load Balancer Name : #{load_balancer_name} | Subnets : #{subnet_array} | Listeners : #{listener_array}")
if !load_balancer_name.nil? && !load_balancer_name.empty?
		name_validation = @call.connector(connector_name).set('action','describe-application-load-balancer').set('name',load_balancer_name)
		validation_response = name_validation.sync
		if validation_response.exitcode != 0

								if !@input.get('subnet1').nil? && !@input.get('subnet1').empty?
									subnets = []
									subnets << @input.get('subnet1')
										if !@input.get('subnet2').nil? && !@input.get('subnet2').empty?
											subnets << @input.get('subnet2')
											connector_call = @call.connector(connector_name)
									                          .set('action', action)
									                          .set('name',load_balancer_name)
									                          .set('access-key', @access_key)
									                          .set('security-key', @secret_key)
																						.set('subnets',subnets)

																if !request_timeout.nil? && !request_timeout.empty?
																	connector_call.timeout(request_timeout)
																else
																	@log.trace("Calling #{connector_name} with default timeout...")
																end
																if !region.nil? && !region.empty?
														         response = connector_call.set('region', region).sync
														    else
														    	response = connector_call.sync
														        @log.trace("region is not provided so using default region 'us-east-1'")
														    end
										else
											raise "Please provide at least two subnets"
										end

							elsif !subnet_array.nil? && !subnet_array.empty?
								connector_call = @call.connector(connector_name)
								                          .set('action', action)
								                          .set('name',load_balancer_name)
								                          .set('access-key', @access_key)
								                          .set('security-key', @secret_key)
									                      .set('listeners',listener_array)


									if !subnet_array.nil? && !subnet_array.empty?
									    connector_call.set('subnets',subnet_array)
									else
											raise "Subnets is not provided"
									end
									if !scheme.nil? && !scheme.empty?
										connector_call.set('scheme',scheme)
									end

									if !request_timeout.nil? && !request_timeout.empty?
										connector_call.timeout(request_timeout)
									else
										@log.trace("Calling #{connector_name} with default timeout...")
									end

									if !security_groups.nil? && security_groups.empty?
										connector_call.set('security-groups',security_groups)
									end

									if !tags.nil? && tags.empty?
										connector_call.set('tags',tags)
									end
									if !region.nil? && !region.empty?
								         response = connector_call.set('region', region).sync
								    else
								    	response = connector_call.sync
								        @log.trace("region is not provided so using default region 'us-east-1'")
								    end
							else
								raise "Please provide subnets"
							end
		else
				raise "Load balancer with the given name '#{load_balancer_name}' already exists"
		end
else
	@log.error("Error: At 'Load balancer name' #{load_balancer_name}. Please provide load balancer name.")
end

@log.info("RESPONSE OF CREATE LOAD BALANCER>>>>#{response}")

# Amazon EC2 Connector Response Meta Parameters
response_exitcode = response.exitcode	# Exit status code
response_message = response.message	# Execution status messages

if response_exitcode == 0
    @log.info("SUCCESS in executing #{connector_name} where, exitcode : #{response_exitcode} | message : #{response_message}")
    @output.set('exit-code', 0).set('message', response_message)
else
    @log.error("ERROR in executing #{connector_name} where, exitcode : #{response_exitcode} | message : #{response_message}")
		@log.error("ERROR in executing #{connector_name} where, exitcode : #{response.exitcode} | message : #{response.message}")
		response=response.to_s
		if !response.empty?
		@output.set('message', response_message).set('exit-code', 1).setraw('error-details',response.to_s)
		else
		@output.set('message', response_message).set('exit-code', 1)
		end
end
@log.trace("Finished executing 'fb-cloud:aws-ec2:operation:create_application_load_balancer.rb' flintbit")
# end

File: aws-ec2/operation/workflow/create_application_load_balancer.js
/*
 * Creation Date: 28/05/2019
 * Summary: To create application load balancer
 * Description: To create application load balancer AWs using aws-ec2 connector
 */

log.trace("Started executing 'fb-cloud:aws-ec2:operation:workflow:create_application_load_balancer.js'");

input_scope = JSON.parse(input);   //For checking if the input contains the necessary keys

//Connector Name  - mandatory
connector_name = "amazon-ec2";
connector_call = call.connector(connector_name);
log.info("Connector Name: "+connector_name);

//Action - mandatory
action = "create-application-load-balancer";
log.info("Action: "+action);

connector_call.set("action",action);

if(input_scope.hasOwnProperty("cloud_connection")){

    //Access Key - mandatory
    access_key = input_scope.cloud_connection.encryptedCredentials["access_key"];
    if(access_key!=null || access_key!=""){
       connector_call.set("access-key",access_key); 
       log.info("Access Key is given");
    }
    else{
        log.error("Access-Key is null or empty string.");
    }

    //Security Key - mandatory
    secret_key = input_scope.cloud_connection.encryptedCredentials["secret_key"];
    if(secret_key!=null || secret_key!=""){
       connector_call.set("security-key",secret_key); 
       log.info("Security Key is given");
    }
    else{
        log.error("Security-Key is null or empty string.");
    }

    //Load Balancer Name
    if(input_scope.hasOwnProperty("load_balancer_name")){
        load_balancer_name = input.get("load_balancer_name");
        if(load_balancer_name!=null || load_balancer_name!=""){
            connector_call.set("name",load_balancer_name); 
            log.info("Load Balancer Name: "+load_balancer_name);
        }
        else{
            log.error("Load Balancer Name is null or empty string.")
        }
    }
    else{  //load_balancer_name key not present in input JSON 
        log.error("Input does not contain the key 'load_balancer_name'")
    }

    //Subnet ID - mandatory
    //minimum 2 subnets required from different availability zones.
    if(input_scope.hasOwnProperty("subnet_id")){
        subnet_id = input.get("subnet_id");
        index = subnet_id.indexOf(","); //to check that user has entered minimum 2 subnet-IDs
        if(index!=-1){
            subnet_id = subnet_id.split(","); //Splitting into an array
            connector_call.set("subnets",subnet_id); //Passing an array
            log.info("Subnets: "+subnet_id);
        }
        else{
            log.error("subnet_id must contain minimum 2 subnet-IDs.")
        }
    }
    else{  //subnet_id key not present in input JSON 
        log.error("Input does not contain the key 'subnet_id'")
    }

    //Region - mandatory
    if(input_scope.hasOwnProperty("region")){
        region = input.get("region");
        if(region!=null || region!=""){
            connector_call.set("region",region); 
            log.info("Region: "+region);
        }
        else{
            log.error("Region is null or empty string.");
        }
    }
    else{  //region key not present in input JSON 
        log.error("Input does not contain the key 'region'");
    }

    //Security Groups - not mandatory
    if(input_scope.hasOwnProperty("security_groups")){
        security_groups = input.get("security_groups");
        if(security_groups!=null || security_groups!=""){
            index = security_groups.indexOf(",");
            if(index!=-1){
                security_groups = security_groups.split(",");
                connector_call.set("security-groups",security_groups); 
                log.info("security_groups: "+security_groups);
            }
            else{
                security_groups_array = [];
                security_groups_array.push(security_groups);
                connector_call.set("security-groups",security_groups_array); 
                log.info("security_groups: "+security_groups_array);
            }
        }
        else{
            log.info("security_groups is null or empty string.");
        }
    }
    else{  //security_groups key not present in input JSON 
        log.info("Input does not contain the key 'security_groups'");
    }

    //Timeout - NOT mandatory
    if(input_scope.hasOwnProperty("request_timeout")){
        request_timeout = input.get("request_timeout");
        if(request_timeout!=null || request_timeout!=""){
            connector_call.set("timeout",request_timeout); 
            log.info("Request Timeout: "+request_timeout);
        }
        else{
            connector_call.set("timeout",240000); 
            log.info("request_timeout not given. Setting 240000 miliseconds as timeout");
        }
    }
    else{
        connector_call.set("timeout",240000); 
        log.info("request_timeout not given. Setting 240000 miliseconds as timeout");
    }
    
    //Connector call
    response = connector_call.sync();

    //Response meta-parameters
    response_exitcode = response.exitcode();
    response_message = response.message();

    if(response_exitcode==0){
        log.info("Application Load Balancer Created Successfully.");
        user_message = "Application Load Balancer Created Successfully.";
        output.set("exit-code",0)
              .set("user_message",user_message)
              .set("message",response_message);
        log.trace("Finished executing 'fb-cloud:aws-ec2:operation:workflow:create_application_load_balancer.js' successfully.");
    }
    else{
        log.error("Failure in execution, message:"+response_message+" | exitcode:"+response_exitcode);
        output.set("error",response_message).set("exit-code",-1);
        log.trace("finished executing 'fb-cloud:aws-ec2:operation:workflow:create_application_load_balancer.js' with errors")
    }
}
else{ //cloud_connection key not present in input JSON
    log.error("Cloud Connection not given. Can not authenticate without Secret-Key and Access-Key");
}
